### ISSUE TITLE
[HttpFoundation] IpUtils::$checkedIps is never cleared

### ISSUE THREAD

**Issue by danielburger1337** [symfony/symfony#50032]
### Symfony version(s) affected

5.4, 6.2

### Description

Foreword: I don't really know how I can make this bug report more concise while still presenting my entire thought process.
Sorry for the long read!

To avoid duplicate work, the IpUtils utility class caches its IP validation results but **never** clears its cache.

Because Symfony is also supporting "infinite" runtimes like swoole or FrankenPHP, the memory usage of this array may become pretty sizeable if either an application checks many IP Adresses per request or the runner is running for a long time.

I might be missing something because I would be kind of surprised if no one encountered this memory usage issue before. So there might be a way already built in by the runtimes that garbage collect static arrays like this.

The best comparison I can think of is the [ArrayAdapter](https://github.com/symfony/symfony/blob/6.3/src/Symfony/Component/Cache/Adapter/ArrayAdapter.php) from the cache component which would have the same problem if it wouldnt implement the [ResettableInterface](https://github.com/symfony/symfony/blob/6.3/src/Symfony/Component/Cache/ResettableInterface.php) which inherits from [ResetInterface](https://github.com/symfony/symfony/blob/6.3/src/Symfony/Contracts/Service/ResetInterface.php). 

### How to reproduce

I had many issues setting up runtimes like swoole or FrankenPHP back when I tried it, so I kind of try to avoid it right now,
but I am pretty sure that the fine people contributing to this project can get the gist of the potential issue I described above. 
Otherwise Im sorry that I couldnt express my thoughts more clearly.

### Possible Solution

Now, we could register the IpUtils utility as a service in the FrameworkBundle and tag it with "kernel.reset", but I think it would be a huge overkill because all methods of IpUtils are currently static and therefor the extra cost of adding it to the container for literally **every single Symfony application in the world** is just not worth it in my opinion.

Instead I would propose the following two (very similar) options:

1. Add something along the lines of `IpUtils::clearCache` and call it in Kernel::boot (before L113 or after L115)
https://github.com/symfony/symfony/blob/f84279720ae4de0942f8d1b41d6161ee321d1ade/src/Symfony/Component/HttpKernel/Kernel.php#L109-L116
2. Add something along the lines of `IpUtils::clearCache` and call it in the [ServicesResetter](https://github.com/symfony/symfony/blob/6.3/src/Symfony/Component/HttpKernel/DependencyInjection/ServicesResetter.php) directly. This wouldnt be ideal because the service may be overwritten or not registered at all. Both is really rare (I think) but worth the consideration.

Everyone who uses a runtime such as swoole (notably not FrankenPHP) is basically forced to use a reverse proxy and [Request::isFromTrustedProxy](https://github.com/symfony/symfony/blob/f84279720ae4de0942f8d1b41d6161ee321d1ade/src/Symfony/Component/HttpFoundation/Request.php#L1964-L1967) is therefor called on every request (even if they use REMOTE_ADDR as trusted proxy value). 

Somehow not invalidating the cache on every request could be quite benefitial because of this. 
We could implement something like a probabilistic cache invalidation or just clear the cache when a fixed size is reached (we know that only boolean values with string keys are cached). 

Registering IpUtils as a service would definitly be the more "Symfony" way of handling this issue but I think the `Kernel::boot` hack is worth it and **NOT** totally unprecedented since [Kernel::preBoot](https://github.com/symfony/symfony/blob/f84279720ae4de0942f8d1b41d6161ee321d1ade/src/Symfony/Component/HttpKernel/Kernel.php#L745-L770) actually sets the trusted proxies / hosts.

**Comment by stof** [comment]
I suggest implementing a max size in the cache instead.

the ArrayAdapter of the Cache component already has an implementation for that.

**Comment by danielburger1337** [comment]
**TLDR**:
Strict size limit of 1024 and removing about a tenth when cache size limit is reached should have the least perfomance impact.
___

Thats a great idea.
Strict size limit or LRU strategy?

LRU would have a small perfomance penalty since on cache hit the item has to be moved to first position.

Limit of 1024 would probably be a good idea.

Also, when the cache limit is reached (for both strategies), should we remove more than just one cache item at once? 
Otherwise the `array_slice` / `array_shift` operation will be executed on every new cache entry exceeding the size limit which is slower and more memory expensive than just rechecking an IP address.

I would suggest removing a tenth of the size limit when limit was reached (so e.g. 1024 / 10 = ~102 items).

Only removing one cache item has had huge perfomance impacts in my small test:

```php
use Symfony\Component\HttpFoundation\IpUtils;

$start = microtime(true);

for ($i = 0; $i < 1_000_000; ++$i) {
    IpUtils::checkIp4(random_int(1, 254).random_int(0, 254).random_int(0, 254).random_int(0, 255), '192.168.1.1');
}

$end = microtime(true) ;

var_dump($end - $start);

// Removing about 102 items at a time (array_slice): ~2.6 seconds
// Removing 1 item at a time (array_shift): ~4 seconds
// Removing 1 item at a time (array_slice) ~15.1 seconds (no surprise here)
```

These perfomance numbers should be taken with a grain of salt because most applications will never check that many IPs, but its a good indicator of the potential bottleneck.


Should I base my PR on 5.4 since it wont introduce any BC and is not really a new feature?

**Comment by nicolas-grekas** [comment]
This reminds me about https://github.com/symfony/symfony/blob/33da4c8b0775a180483ebd46d878b50cc414072f/src/Symfony/Component/Cache/Traits/AbstractAdapterTrait.php#L349-L353

And https://github.com/symfony/symfony/blob/33da4c8b0775a180483ebd46d878b50cc414072f/src/Symfony/Component/Cache/Adapter/ArrayAdapter.php?rgh-link-date=2023-04-16T20%3A03%3A10Z#L282-L287

This should be for 6.3 as that's not a bugfix but an improvement to me.

**Comment by stof** [comment]
> Should I base my PR on 5.4 since it wont introduce any BC and is not really a new feature?

The rule is not "if it is not a feature it goes to older branch". It is "if it is not a bugfix, it goes to latest branch".

**Comment by danielburger1337** [comment]
> The rule is not "if it is not a feature it goes to older branch". It is "if it is not a bugfix, it goes to latest branch".

Normally I would 100% agree, but since this is a fix of a potential memory leak (even though the possibility that it actually happens is really really low), I wasnt quite so sure whether it counts as a bug fix or just a general improvement.

Anyway, I will submit my PR against 6.3
